bookshelf/api_v3/rackspace.py

```diff
@@ -0,0 +1,231 @@
+    auth_system = field(factory=unicode, mandatory=True)
+    auth_auth_url = field(factory=unicode, mandatory=True)
+    tenant = field(factory=unicode, mandatory=True)
+    ami = field(factory=unicode, mandatory=True)
+    description = field(factory=unicode, mandatory=True)
+    instance_name = field(factory=unicode, mandatory=True)
+
+
+class RackspaceState(PClass):
+    instance_name = field(factory=unicode, mandatory=True)
+    ip_address = field(factory=unicode, mandatory=True)
+    distro = field(factory=unicode, mandatory=True)
+    region = field(factory=unicode, mandatory=True)
+
+
+@implementer(ICloudInstance)
+@provider(ICloudInstanceFactory)
+class Rackspace(object):
+
+    cloud_type = 'rackspace'
+
+    def __init__(self, config, state):
+        self.config = RackspaceConfiguration.create(config)
+        self.state = state
+        self._nova = self._connect_to_rackspace()
+
+    @property
+    def distro(self):
+        return Distribution(self.state.distro)
+
+    @property
+    def username(self):
+        return self.config.username
+
+    @property
+    def description(self):
+        return self.config.description
+
+    @property
+    def ip_address(self):
+        return self.state.ip_address
+
+    @property
+    def image_basename(self):
+        return self.config.image_basename
+
+    @property
+    def region(self):
+        return self.state.region
+
+    @property
+    def key_filename(self):
+        return self.config.private_key_filename
+
+    @classmethod
+    def create_from_config(cls, config, distro, region):
+        distro = distro.value
+        instance_name = "{}-{}".format(
+            config['instance_name'],
+            unicode(uuid.uuid4())
+        )
+        state = RackspaceState(
+            instance_name=instance_name,
+            ip_address="",
+            distro=distro,
+            region=region
+        )
+        instance = Rackspace(config, state)
+        instance.upload_key()
+        instance._create_server()
+        return instance
+
+    def upload_key(self):
+
+        try:
+            log_green("Checking for key pair {}".format(self.config.key_pair))
+            self._nova.keypairs.get(self.config.key_pair)
+            log_green("Key pair exists in rackspace")
+        except NotFound:
+            log_green("Creating key pair {}".format(self.config.key_pair))
+            with open(self.config.public_key_filename) as keyfile:
+                self._nova.keypairs.create(self.config.key_pair,
+                                           keyfile.read())
+
+    @classmethod
+    def create_from_saved_state(cls, config, saved_state):
+        state = RackspaceState.create(saved_state)
+        instance = Rackspace(config, state)
+        server = instance._ensure_instance_running()
+        # if we've restarted a terminated server, the ip address
+        # might have changed from our saved state, get the
+        # networking info and resave the state
```

Comment:
is there value in doing this in Rackspace?
as starting a terminated server will always create a new server, since shutdown a box in rackspace always means destroy this box